Services/Models/InternalDeviceState.cs

```diff
@@ -0,0 +1,247 @@
+        public void SetState(Dictionary<string, object> newState)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                this.simulationState = newState;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if key exists
+        /// </summary>
+        public bool HasStateValue(string key)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                return this.simulationState.ContainsKey(key);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve simulated device state value.
+        /// </summary>
+        public object GetStateValue(string key)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                return this.simulationState[key];
+            }
+        }
+
+        /// <summary>
+        /// Set the simulation state value with the given key, adds new value if key does not exist.
+        /// </summary>
+        public void SetStateValue(string key, object value)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                if (this.simulationState.ContainsKey(key))
+                {
+                    this.simulationState[key] = value;
+                }
+                else
+                {
+                    this.simulationState.Add(key, value);
+                }
+            }
+        }
+
+        private Dictionary<string, object> SetupTelemetry(DeviceModel deviceModel)
+        {
+            // put telemetry properties in state
+            Dictionary<string, object> state = CloneObject(deviceModel.Simulation.InitialState);
+
+            // Ensure the state contains the "online" key
+            if (!state.ContainsKey("online"))
+            {
+                state["online"] = true;
+            }
+
+            // TODO:This is used to control whether telemetry is calculated in UpdateDeviceState.
+            // methods can turn telemetry off/on; e.g. setting temp high- turnoff, set low, turn on
+            // it would be better to do this at the telemetry item level - we should add this in the future
+            state.Add(CALC_TELEMETRY, true);
+
+            return state;
+        }
+
+        /// <summary>
+        /// Initializes device twin with initial device twin properties from the device model.
+        /// </summary>
+        private Dictionary<string, object> SetupProperties(DeviceModel deviceModel)
+        {
+
+            Dictionary<string, object> result = new Dictionary<string, object>();
+
+            // add device properties to the properties dictionary. These properties will be written
```

Comment:
hypernit: please start comments with capital letter. About the rest of the comment, I'm not sure if we should call out the simulation behavior in here. Anyhow, I would partially rephrase saying "These properties will be stored as 'reported properties' in the 'device twin' in Azure IoT Hub"